### DISCUSSION TITLE
How to create environment file variable programmatically (using Java)

### DISCUSSION THREAD

**Discussion by alaa-ayyad** [community/community#36839] [Actions] [answered]
I started replacing `set-output` my with `GITHUB_OUTPUT` environment file after the announcement of [Deprecating save-state and set-output commands](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/).

As part of my workflow, I have Java class that creates and adds a variable to GITHUB workflow context like this:
`System.out.println("::set-output name=test_passed::" + result);`

Which works perfectly and I can even use this variable in the following Workflow steps, but now I need to switch to environment files and I couldn't write the Java equivalent to this:
```
- name: Set output
  run: echo "{name}={value}" >> $GITHUB_OUTPUT
```

More info: This Java class runs during `mvn verify` step, it reads some xml and json files from Serenity BDD report and creates a flag `test_passed` that can be used as a condition in the following Workflow steps, thus it's vital to create this variable from Java.

Is it possible to add variables to `$GITHUB_OUTPUT` from Java?

**Comment by antonblr** [comment, marked as answer]
Not a Java expert, but essentially you just need to append a string to the `$GITHUB_OUTPUT` file.

```java
String filePath = System.getenv("GITHUB_OUTPUT");
void appendToFile(String filePath, String content) {
  try (FileWriter fw = new FileWriter(filePath, true)) {
    fw.write(content + System.lineSeparator());
  } catch (IOException e) { 
    // TODO handle exception
  }
}
```

**Reply by alaa-ayyad** [reply]
@antonblr Your solution worked, thank you! It is good reminder that `GITHUB_OUTPUT` is a file!

I was able to store new value to `$GITHUB_OUTPUT` from my Java class and then display it in the consequent Github Workflow step:
![Screenshot 2022-10-21 at 00-15-17 Integration Tests · telus_id-cust-ci-credential-management@1810ad0](https://user-images.githubusercontent.com/12691011/197136967-672a6e4b-ec6a-40d2-8573-d56c735e1f38.png)
And of course, variable is referenced in the Workflow as your normally do with any step output: `${{ steps.<maven step id>.outputs.<var name>}}`

I'm sure this solution will come handy to many developers, just be aware that variable should be stored in the format:
`{name}={value}` 

Thus I made this slight modification to your code:
```
import java.io.FileWriter;
import java.io.IOException;

public class GithubEnv {
    
  private static final String filePath = System.getenv("GITHUB_OUTPUT");
  
  public static void setVariable(String name, String value) {
    String content = name + "=" + value;
    try (FileWriter fw = new FileWriter(filePath, true)) {
      fw.write(content + System.lineSeparator());
    } catch (IOException e) { 
      // TODO handle exception
    }
  }
}
```
And then utlity can be used like this:
`GithubEnv.setVariable( "total_tests", total );`

Cheers!

**Comment by ossanna16** [comment]
Thank you for asking a good question @alaa-ayyad, I'm glad you found our community 🙂
